org.spoofax.terms/src/org/spoofax/interpreter/terms/IStrategoArrayList.java

```diff
@@ -0,0 +1,234 @@
+    }
+
+    static StrategoArrayListBuilder arrayListBuilder() {
+        return new StrategoArrayListBuilder(16);
+    }
+
+    static StrategoArrayListBuilder arrayListBuilder(int size) {
+        return new StrategoArrayListBuilder(size);
+    }
+
+    @Override
+    default boolean internalDoSlowMatch(IStrategoTerm second) {
+        if(this == second) {
+            return true;
+        }
+        if(!TermUtils.isList(second)) {
+            return false;
+        }
+        if(this.getSubtermCount() != second.getSubtermCount()) {
+            return false;
+        }
+
+        if(second instanceof StrategoArrayList) {
+            IStrategoArrayList other = (IStrategoArrayList) second;
+
+            if(!this.getAnnotations().match(other.getAnnotations())) {
+                return false;
+            }
+
+            //noinspection ArrayEquality
+            if(this.internalGetBackingArray() == other.internalGetBackingArray()) {
+                return internalGetOffset() == other.internalGetOffset() && this.internalGetEndOffset() == other.internalGetEndOffset();
+            }
+
+            Iterator<IStrategoTerm> termsThis = this.iterator();
+            Iterator<IStrategoTerm> termsOther = other.iterator();
+
+            if(!this.isEmpty()) {
+                for(IStrategoTerm thisNext = termsThis.next(), otherNext = termsOther.next()
+                    ; termsThis.hasNext()
+                        ; thisNext = termsThis.next(), otherNext = termsOther.next()) {
+                    if(thisNext != otherNext && !thisNext.match(otherNext)) {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        final IStrategoList snd = (IStrategoList) second;
+
+        if(!isEmpty()) {
+            IStrategoTerm head = head();
+            IStrategoTerm head2 = snd.head();
+            if(head != head2 && !head.match(head2))
+                return false;
+
+            IStrategoList tail = tail();
+            IStrategoList tail2 = snd.tail();
+
+            for(IStrategoList cons = tail, cons2 = tail2; !cons.isEmpty(); cons = cons.tail(), cons2 = cons2.tail()) {
+                IStrategoTerm consHead = cons.head();
+                IStrategoTerm cons2Head = cons2.head();
+                if(!cons.getAnnotations().match(cons2.getAnnotations())) {
+                    return false;
+                }
+                if(consHead != cons2Head && !consHead.match(cons2Head))
+                    return false;
+            }
+        }
+
+        IStrategoList annotations = getAnnotations();
+        IStrategoList secondAnnotations = second.getAnnotations();
+        if(annotations == secondAnnotations) {
+            return true;
+        } else
+            return annotations.match(secondAnnotations);
+    }
+
+    /**
+     * N.B. this implementation may look strange but it's designed to return the same hashcode as {@link StrategoList#hashFunction()}
```

Comment:
Instead of duplicating the code and risk it going out of sync, can you extract this into a static method or some other way to share the code?

Reply:
It's possible to pull it up into `IStrategoList`, but it's actually required by `StrategoTerm`, which is somewhere else in the type hierarchy, so it's a bit awkward. Still probably a good idea though.